lib/src/revset.rs

```diff
@@ -2194,7 +2216,24 @@ impl SymbolResolver for DefaultSymbolResolver<'_> {
             }
         }
 
-        Err(make_no_such_symbol_error(self.repo, symbol))
+        // Let an extension resolve the symbol.
+        let ext_ids = self.extensions.iter().try_fold(
+            None,
+            |ids, r| -> Result<Option<Vec<CommitId>>, RevsetResolutionError> {
+                if let Some(new_ids) = r.resolve_symbol(symbol)? {
+                    match ids {
+                        Some(prev_ids) if !ids_equal(&prev_ids, &new_ids) => {
+                            Err(RevsetResolutionError::AmbiguousSymbol(symbol.to_owned()))
+                        }
+                        _ => Ok(Some(new_ids)),
+                    }
+                } else {
+                    Ok(ids)
+                }
+            },
+        )?;
```

Comment:
Does it make sense to disambiguate symbols between separate namespaces?

My understanding is that `PartialSymbolResolver` represents a namespace like "tags" or "branches", and these namespaces have predefined precedence order. We can of course make it an error, but we'll need a better error indication to suggest `tags(name)` for example.

Reply:
The intent here is that `PartialSymbolResolver` is per extension, not per namespace.  If an extension defines multiple namespaces with precedence order they should all be handled within one PartialSymbolResolver.  If multiple extensions try to resolve the same symbol we should only allow it if they agree on what the resolution should be (which seems unlikely to happy in practice, but unlike other extension points like factories, we can at least determine that they are 'the same' here and be nicer to multiple extensions).

I added comments on the traits to clarify.

Reply:
> Is the error actionable by user?

In this case I think 'yes', because the symbol is user input.  The user can take action by either providing a different input, or by changing their set of installed extensions.

It's the same problem as two extensions both trying to register the same working copy factory or template function - there we panic, because we detect the problem at init() time.  Here we can't detect it until symbol resolution time but it's the same conceptual issue.  I think erroring is appropriate so that users don't get silently wrong behavior from conflicting extensions (e.g., I install an extension that handles symbol "foo", and run instructions that reference this, but they silently do the wrong thing because some other extension it doesn't know about also handles "foo")

Reply:
> > Is the error actionable by user?
> 
> In this case I think 'yes', because the symbol is user input. The user can take action by either providing a different input, or by changing their set of installed extensions.
> 
> It's the same problem as two extensions both trying to register the same working copy factory or template function - there we panic, because we detect the problem at init() time. Here we can't detect it until symbol resolution time but it's the same conceptual issue. I think erroring is appropriate so that users don't get silently wrong behavior from conflicting extensions (e.g., I install an extension that handles symbol "foo", and run instructions that reference this, but they silently do the wrong thing because some other extension it doesn't know about also handles "foo")

Hmm, from my perspective, the symbol "foo" is ambiguous for the reason the user doesn't know. If two extensions resolve "foo" differently, it implies that they look up the symbol in different namespaces, and one of them should be prioritized (like "tags" vs "branches".) If it's a bug (like factory name conflicts), symbol resolution could panic there, but I don't think it's reasonable to assume that a certain symbol is reserved to one extension.

Reply:
I don't think it's accurate to say the user doesn't know the things the extensions they have support.  I'd really like to avoid having one extension 'win' over the others because it introduces a behavioral dependency on the _order_ in which extensions are loaded, something which has caused endless pain for Mercurial setups.  So far with the map-key-based extension merging, we've remained order-agnostic and I think it's valuable to preserve that.

Reply:
> I'd really like to avoid having one extension 'win' over the others because it introduces a behavioral dependency on the _order_ in which extensions are loaded, something which has caused endless pain for Mercurial setups.

I agree about that, but revset symbol resolution inherently has this problem because there are multiple independent namespaces in which the symbol will be looked up. This is quite different from e.g. template keywords or revset functions.

I think we'll eventually need some mechanism to overlay `PartialSymbolResolver` in deterministic order (e.g. tag-like extension symbols should precede builtin branches lookup), but my feeling is that we aren't reached to that point yet.

Reply:
> but my feeling is that we aren't reached to that point yet.

So is there anything actionable now other than a TODO?  I'll leave that to help document future intentions but for now I'd rather not define the API we're not ready for yet.

Reply:
> So is there anything actionable now other than a TODO? I'll leave that to help document future intentions but for now I'd rather not define the API we're not ready for yet.

I think TODO comment is pretty okay. I don't think the extension API is widely used, so we can freely change it.

Alternatively, if it makes sense to disambiguate symbols within multiple namespaces/extensions, maybe we can migrate tags/branches/git_refs resolution first? I don't know if that's better than the current precedence-based resolution, but we can remove inconsistency between core symbol and extension symbol resolution.